Add abseil cgmanifest declaration. Update coding standards. (#10374)
microsoft/onnxruntime: docs/Coding_Conventions_and_Standards.md

* When adding a new class, disable copy/assignment/move until you have a proven need for these capabilities. If a need arises, enable copy/assignment/move selectively, and when doing so validate that the implementation of the class supports what is being enabled.
  * Use ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE initially
  * See the other ORT_DISALLOW_* macros in https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/common/common.h
* Consider using 'const gsl::span<const T>&' (or 'std::span' when supported) as input arguments when passing const references to containers with contiguous storage (like 'std::vector'). This allows to make the function container independent, represent arbitrary memory spans or pass sub-spans as an argument.
* The use of the following container typedefs to reduce memory allocations is preferred:

>   * Use 'TensorShapeVector' typedef to build or modify shapes from core/framework/tensor_shape.h. It is based on a vector implementation that features small buffer optimization. Its small buffer size is the same to that of in TensorShape. Use 'InlinedShapeVector<T>'    for shape related operations, but of different type.

Reviewer: ```suggestion
  * Use 'TensorShapeVector' typedef to build or modify shapes from core/framework/tensor_shape.h. It is based on a vector implementation that features small buffer optimization. Its small buffer size is the same to that of in TensorShape. Use 'InlinedShapeVector<T>' for shape related operations, but of different type.
```

>   * Use 'TensorShapeVector' typedef to build or modify shapes from core/framework/tensor_shape.h. It is based on a vector implementation that features small buffer optimization. Its small buffer size is the same to that of in TensorShape. Use 'InlinedShapeVectorT<T>'    for shape related operations, but of different type.

Reviewer 2: ```suggestion
  * Use 'TensorShapeVector' typedef to build or modify shapes from core/framework/tensor_shape.h. It is based on a vector implementation that features small buffer optimization. Its small buffer size is the same to that of in TensorShape. Use 'InlinedShapeVector<T>' for shape related operations, but of different type.
```

  * Use 'InlinedVector<T>' typedef instead of std::vector. By default, it provides 64 bytes of inlined storage. You can customize inlined size with the second template non-type parameter N.
  * Use 'InlinedHashSet<T>' and 'InlinedHashMap<T>' typedefs from core/framework/inlined_containers.h. These are drop-in replacements for 'std::unordered_set/map' that store their keys and values in one continuous buffer and reduce the number of allocations. They also do not allocate an 'end' node. Note, that these Hash containers do not provide pointer stability.
  * Consider using 'std::string_view' to use in maps and sets to reduce the number of allocations and avoid string duplication. Keep in mind that the strings referred to must be alive.
  * We have selected to use Abseil library for the above typedefs. Abseil container documentation is [here](https://abseil.io/docs/cpp/guides/container#abseil-containers).

> * Prefer using `reserve()` and not `resize()` on vectors.

Reviewer: perhaps more explicitly mention the possibly wasteful default construction of resize()? so folks understand the motivation for this.

Author: Removed

> * Prefer using `reserve()` and not `resize()` on vectors. 'resize()' default constructs all the elements for the size which can be expensive/noticible even if the type is trivial. Default values are rarely used in practice and it becomes a waste. Construction like 'std::vector<int>(10, 0)' is the same as 'resize()' and is potentially wasteful.

Reviewer 2: ```suggestion
* Prefer using `reserve()` and not `resize()` on vectors. 'resize()' default constructs all the elements for the size which can be expensive/noticeable even if the type is trivial. Default values are rarely used in practice and it becomes a waste. Construction like 'std::vector<int>(10, 0)' is the same as 'resize()' and is potentially wasteful.
```

> * Prefer using `reserve()` and not `resize()` on vectors. 'resize()' default constructs all the elements for the size which can be expensive/noticiable even if the type is trivial. Default values are rarely used in practice and it becomes a waste. Construction like 'std::vector<int>(10, 0)' is the same as 'resize()' and is potentially wasteful.

Reviewer: ```suggestion
* Prefer using `reserve()` and not `resize()` on vectors. 'resize()' default constructs all the elements for the size which can be expensive/noticeable even if the type is trivial. Default values are rarely used in practice and it becomes a waste. Construction like 'std::vector<int>(10, 0)' is the same as 'resize()' and is potentially wasteful.
```

* When adding a new class, disable copy/assignment/move until you have a proven need for these capabilities. If a need arises, enable copy/assignment/move selectively, and when doing so validate that the implementation of the class supports what is being enabled.
  * Use ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE initially
  * See the other ORT_DISALLOW_* macros in https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/common/common.h

> * Consider using const gsl::span<const T>& (or std::span when supported) as input arguments when passing const references to vector like containers. This allows to make the function container independent, represent arbitrary memory spans or pass sub-spans into it.

Reviewer: ```suggestion
* Consider using const gsl::span<const T>& (or std::span when supported) as input arguments when passing const references to containers with contiguous storage (like std::vector). This allows to make the function container independent, represent arbitrary memory spans or pass sub-spans into it.
```

Author: Re-worded

* Use TensorShapeVector to build or modify shapes from core/framework/tensor_shape.h. It is based on a vector implementation that features small buffer optimization. Its small buffer size is the same that we use in TensorShape. Use InlinedShapeVectorT<T> for shape related operations, but of different type.

> * Use InlinedHashSet<T> and InlinedHashMap<T> from core/framework/inlined_containers.h. These are drop-in replacements for std::unordered_set/map that store their keys and values in one continuous buffer and reduce the number of frequency allocations. They also do not allocate an 'end' node. Note, that these Hash containers do not provide pointer stability. Use InlinedVector<T, N> when there is a reason to believe your code may benefit from small buffer optimization and you have a good idea (data) what size it must be.

Reviewer 3: Looking at the usage of InlinedShapeVectorT<T> it seems like it could be used for operations on small vectors of a different type which may or may not related to the shape. 

Based on that could it be renamed something more generic and the text here simply say that it's for usage with vectors with less than X number of elements?

Author: Introduced default N calculation for InlinedVector<T>, it is now suitable for general use.

> * Use InlinedHashSet<T> and InlinedHashMap<T> from core/framework/inlined_containers.h. These are drop-in replacements for std::unordered_set/map that store their keys and values in one continuous buffer and reduce the number of frequency allocations. They also do not allocate an 'end' node. Note, that these Hash containers do not provide pointer stability. Use InlinedVector<T, N> when there is a reason to believe your code may benefit from small buffer optimization and you have a good idea (data) what size it must be.

Reviewer 2: ```suggestion
* Use InlinedHashSet<T> and InlinedHashMap<T> from core/framework/inlined_containers.h. These are drop-in replacements for std::unordered_set/map that store their keys and values in one continuous buffer and reduce the number of frequent per-node allocations. They also do not allocate an 'end' node. Note, that these Hash containers do not provide pointer stability. Use InlinedVector<T, N> when there is a reason to believe your code may benefit from small buffer optimization and you have a good idea (data) what size it must be.
```

Can also point to the https://abseil.io/docs/cpp/guides/container if someone wants to read more.

Author: Done.

> * Use InlinedHashSet<T> and InlinedHashMap<T> from core/framework/inlined_containers.h. These are drop-in replacements for std::unordered_set/map that store their keys and values in one continuous buffer and reduce the number of frequency allocations. They also do not allocate an 'end' node. Note, that these Hash containers do not provide pointer stability. Use InlinedVector<T, N> when there is a reason to believe your code may benefit from small buffer optimization and you have a good idea (data) what size it must be.

Reviewer: Is there a suggestion for what the default container choice is? Perhaps I may not have a good idea (data) of the expected size yet. Should I still reach for `std::vector<T>`, or perhaps `InlinedVector<T, kDefaultInlinedVectorSmallBufferSize>`?

Author: Introduced default inlined size calculation so InlinedVector<T> is now suitable for general use.

> * Don't use else after return. see: [https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return](https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return)

Reviewer 3: not sure what 'number of frequency allocations' should be. 

maybe a new bulletpoint ' for 'Use InlinedVector<T, N>...'

Author: Re-worded
